kipoiseq/dataloaders/prototype.py

```diff
@@ -0,0 +1,291 @@
+
+class VariantSequenceExtractor:
+    
+    def __init__(
+        self,
+        reference_sequence_extractor : BaseExtractor,
+        anchor = 0,
+        fixed_len = True
+    ):
+        self.anchor = anchor
+        self.fixed_len = fixed_len
+        self.extractor = VariantSeqExtractor(
+            reference_sequence = reference_sequence_extractor
+        )
+        
+    def extract(
+        self,
+        interval, 
+        variants
+    ):
+        return self.extractor.extract(interval, variants, 
+                anchor=self.anchor, fixed_len=self.fixed_len)
+
+### The Generic Dataloader ###
+        
+class GenericVariantDataloader(SampleIterator):
+    
+    def __init__(
+        self,
+        interval_fetcher,
+        variant_source,
+        variant_insertion_strategy,
+        reference_sequence_extractor,
+        variant_sequence_extractor,
+        sequence_transformer,
+        interval_attrs,
+        anchor = 0,
+        fixed_len = False
+    ):
+        # Generic variant dataloader ingredients:
+        # A interval source (called as iterator)
+        self.interval_fetcher = interval_fetcher.items()
+        # A variant source
+        self.variant_source = variant_source
+        #A variant insertion strategy
+        self.variant_insertion_strategy = variant_insertion_strategy
+        # A sequence source
+        self.reference_sequence_extractor = reference_sequence_extractor
+        # A variant sequence extractor
+        self.variant_sequence_extractor = variant_sequence_extractor
+        # A sequence transformer object (could be identity transform)
+        self.sequence_transformer = sequence_transformer
+        # Build iterator object
+        self.generator = self._generator()
+        # interval attributes
+        self.interval_attrs = interval_attrs
+    
+    def _build_sequences(self, alt_seq_template):
+        ref_seq_parts = [self.reference_sequence_extractor.extract(interval) 
+                        for interval, variants in alt_seq_template]
+        alt_seq_parts = []
+        for idx, (interval, variants) in enumerate(alt_seq_template):
+            if len(variants) > 0:
+                alt_seq_parts.append(self.variant_sequence_extractor.extract(interval, variants))
+            else:
+                alt_seq_parts.append(ref_seq_parts[idx])
+        if self.reference_sequence_extractor.use_strand and alt_seq_template[0][0].strand == "-":
+            ref_seq_parts.reverse()
+            alt_seq_parts.reverse()
+        ref_seq = "".join(ref_seq_parts)
+        alt_seq = "".join(alt_seq_parts)
+        return ref_seq, alt_seq
+            
+    def _generator(self):
+        # Iterate through the interval source
+        for key, intervals in self.interval_fetcher:
+            intervals.sort(key=lambda x: x.start)
+            # Fetch variants corresponding to the intervals
+            intervals_with_variants = [(interval, self.variant_source.fetch_variants(interval))
```

Comment:
This is a very slow way to match variants. A better approach will be overlapping with pyranges if the variant strategy is single variant.

https://github.com/kipoi/kipoiseq/blob/master/kipoiseq/extractors/vcf_matching.py#L148